### PR TITLE
feat(scaletest): add option to send traffic over SSH

### LINKED ISSUE #8242: Scale testing: create utility to generate SSH traffic
# Problem

We currently don't have an easy way to generate SSH traffic to workspaces. The current incarnation of the `scaletest` command runs `ReconnectingPTY` traffic only. SSH traffic over tailnet should, in theory, have a much lower overhead. We want to exercise this at scale however.

# Solution

Create a `coder scaletest` command (or modify an existing command) that will
- Connect to a workspace over SSH
- Run a command
- Assert expected output, if provided

### PR DESCRIPTION
**Pull request by johnstcn**
- Refactors the metrics logic to avoid needing to pass in a whole prometheus registry
- Adds an `--ssh` option to the workspace-traffic command to send SSH traffic
- ~Adds an extra 1 second read-wait after stopping writes. I found that without this, we would hang on closing the connection.~ (removed after PR review)

Fixes https://github.com/coder/coder/issues/8242

### CODE REVIEW

File: scaletest/workspacetraffic/conn.go [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+	if err != nil {
+		_ = agentConn.Close()
+		return nil, xerrors.Errorf("new ssh session: %w", err)
+	}
+	wrappedConn := &wrappedSSHConn{ctx: ctx}
+	wrappedConn.stdout, sshSession.Stdout = io.Pipe()
+	sshSession.Stdin, wrappedConn.stdin = io.Pipe()
+	err = sshSession.RequestPty("xterm", 25, 80, gossh.TerminalModes{})
+	if err != nil {
+		_ = sshSession.Close()
```
Comment (mafredri) [inline scaletest/workspacetraffic/conn.go:52]: Do we need to close the pipes here too?

File: scaletest/workspacetraffic/run.go [hunk 2]
```diff
@@ -83,63 +80,80 @@ func (r *Runner) Run(ctx context.Context, _ string, logs io.Writer) error {
...
-	if rErr := <-rch; rErr != nil {
-		return xerrors.Errorf("read from pty: %w", rErr)
+	// Read for up to one more second.
+	readCtx, readCancel := context.WithTimeout(ctx, time.Second)
+	defer readCancel()
+	select {
+	case <-readCtx.Done():
+		logger.Warn(readCtx, "timed out reading from agent", slog.F("agent_id", agentID))
+	default:
+		rErr := <-rch
```
Comment (mafredri) [inline scaletest/workspacetraffic/run.go:152]: Did you mean this?

```suggestion
	case rErr := <-rch:
```

**Review by mafredri — APPROVED** [review]

**Review by mtojek — APPROVED** [review]
I haven't found anything blocking 👍

File: scaletest/workspacetraffic/conn.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+	})
+	return w.closeErr
+}
+
+func (w *wrappedSSHConn) Read(p []byte) (n int, err error) {
+	select {
+	case <-w.ctx.Done():
+		return 0, xerrors.Errorf("read: %w", w.ctx.Err())
+	default:
+		return w.stdout.Read(p)
```
[the remark this answers is not in the document]
Comment (johnstcn) [reply scaletest/workspacetraffic/conn.go:101]: I ended up just doing this instead:
```
	go func() {
		<-ctx.Done()
		_ = conn.Close()
		close(rch)
	}()
```

**Pull request merged**